src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -130,6 +138,66 @@ static TimestampWrapper validateAndParseTimestampNtzSb16(
     }
   }
 
+  /**
+   * Given a SFTimestamp, get the number of nanoseconds since the last whole second. This
+   * corresponds to the fraction component for Timestamp types
+   *
+   * @param input SFTimestamp
+   * @return Timestamp fraction value, the number of nanoseconds since the last whole second
+   */
+  private static int getFractionFromTimestamp(SFTimestamp input) {
+    BigDecimal epochSecondsInNanoSeconds =
+        new BigDecimal(input.getSeconds().multiply(BigInteger.valueOf(Power10.intTable[9])));
+    return input.getNanosSinceEpoch().subtract(epochSecondsInNanoSeconds).intValue();
+  }
+
+  /**
+   * Validates and parses input for TIMESTAMP_TZ Snowflake type
+   *
+   * @param input TIMESTAMP_TZ in "2021-01-01 01:00:00 +0100" format
+   * @param metadata Column metadata containing scale
+   * @return TimestampWrapper with epoch seconds, fractional seconds, and epoch time in the column
+   *     scale
+   */
+  static TimestampWrapper validateAndParseTimestampTz(Object input, Map<String, String> metadata) {
+    try {
+      if (input instanceof String) {
+        String stringInput = (String) input;
+        int scale = Integer.parseInt(metadata.get(ArrowRowBuffer.COLUMN_SCALE));
+        SFTimestamp timestamp = snowflakeDateTimeFormatter.parse(stringInput);
+        if (timestamp == null) {
+          throw new SFException(
+              ErrorCode.INVALID_ROW, input, "Cannot be converted to a time value");
+        }
+
+        int fraction = getFractionFromTimestamp(timestamp);
+
+        BigDecimal epochTimeAtSecondsScale =
+            timestamp.getNanosSinceEpoch().divide(BigDecimal.valueOf(Power10.intTable[9]));
+        if (fraction % Power10.intTable[9 - scale] != 0) {
+          throw new SFException(
+              ErrorCode.INVALID_ROW,
+              input.toString(),
+              String.format(
+                  "Row specifies accuracy greater than column scale.  fraction=%d, scale=%d",
+                  fraction, scale));
+        }
+        TimestampWrapper wrapper =
+            new TimestampWrapper(
+                timestamp.getSeconds().longValue(),
+                fraction,
+                getTimeInScale(getStringValue(epochTimeAtSecondsScale), scale),
+                timestamp.getTimeZoneOffsetMillis(),
+                timestamp);
+        return wrapper;
+      } else {
+        throw new SFException(ErrorCode.INVALID_ROW, input, "Cannot be converted to a time value");
+      }
```

Comment:
Nit, would it be cleaner to change the above condition to throw if not validate and remove the else?